keps/sig-api-machinery/5366-graceful-leader-transition/README.md

```diff
@@ -0,0 +1,870 @@
+- No Graceful Shutdown: The immediate call to `os.Exit()` prevents any graaceful
+  shutdown or cleanup operations.
+
+
+### Goals
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+- Allow leader-elected components to transition to a follower state without restarting the process
+- Enable graceful handling of leader lock loss
+
+### Non-Goals
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+- This KEP allows work towards graceful shutdown of controllers, but the actual
+  mechanisms of how we do graceful shutdown is outside the scope of this KEP.
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+The main control loops for `kube-controller-manager`, `kube-sceduler`, and
+`cloud-controller-manager` will be updated to support graceful leader
+transitions. When a leader fails to renew its lease, instead of exiting, the
+component will rely on client-go's leader election mechanism to cancel the
+context, and stop its internal controllers. It will then immediately return to a
+follower state where it will attempt to reacquire the lease.
+
+This change will be guarded by a new feature gate, `GracefulLeaderTransition`.
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+Cloud provider A wants to balance controllers over multiple control planes. With
+graceful transitions, multiple locks can be used by a single KCM instance such
+that a subset of components run under each lock.
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+Kubernetes leader elected controllers and the scheduler have been running
+without graceful shutdown for years. There is a risk that controllers and the
+sceduler are not properly respecting context shutdowns. Furthermore, memory
+leaks may exist in the processes that were previously masked by doing a full
+shutdown and restart loop. We will audit each controller carefully but as a
+failstop, the feature will be guarded behind a feature gate. 
```

Comment:
+100. This is the exact risk I'm most concerned with. I need to be convinced we've sufficiently de-risked this before this feature is on-by-default.

Auditing controllers is a necessary step, but I'm not yet convinced it's sufficient.  How would we do the audit? Why do we think the audit process is sufficiently water tight to ensure nothing slips by it?  Are there other tests we should be doing to improve our confidence?

Reply:
There is also a risk that I think we should call out separately:

Say we prove, with sufficiently high confidence, that all the existing controllers shutdown gracefully.  That still leaves this problem:

At any time in the future, a change to an existing controller, an addition of a new controller, or a change to controller infrastructure could create a subtle problem in controller shutdown.  Once we change controller managers to shutdown gracefully, we make the system vulnerable to such mistakes.  How do we detect these?  If we don't , we risk breaking production clusters.